### ISSUE TITLE
gRPC multiple languages on ARM

### ISSUE THREAD

**Issue by tiolan** [grpc/grpc#23336]
### Is your feature request related to a problem? 
I am evaluating gRPC for a large project, running on ARM devices.

As a POC I have a C# project, that uses GRPC C-Core like this:

> <ItemGroup>
    <PackageReference Include="Google.Protobuf" Version="3.12.0" />
    <PackageReference Include="Grpc" Version="2.29.0" />
    <PackageReference Include="Grpc.Tools" Version="2.29.0" PrivateAssets="All" />
  </ItemGroup>

> <ItemGroup>
    <Protobuf Include="../../proto/Protofile1.proto" />
    <Protobuf Include="../../proto/Protofile2.proto" />
    <Protobuf Include="../../proto/Protofile3.proto" />
  </ItemGroup>

When I build on amd64, everything works perfectly.

I now want to compile exactly the same project for ARM on my arm64.
If just executing dotnet restore, I get

> Google.Protobuf.Tools cannot determine host OS and CPU.

So I specify

> ENV PROTOBUF_TOOLS_OS=linux
ENV PROTOBUF_TOOLS_CPU=x64

and get

> The specified task executable "/root/.nuget/packages/grpc.tools/2.29.0/tools/linux_x64/protoc" Exec format error

when building.
(Note that I am building inside Docker - on the arm64 or cross on my amd64, its the same)

So I run the build container manually and did

> dotnet add myproj package Google.Protobuf
dotnet add myproj package Grpc
dotnet add myproj package Grpc.Tools

which provides me the protoc inside the container.
When trying to execute this protoc manually, it says "Exec format error". So it seems, that not an ARM version has been pulled.

Doing the same on the Raspberry directly without Docker has the same effect.

**Back to GRPC:**
As a workaround I generated the source code from the .proto manually, not during the build.

The build then succeeds and I can start the C# app, but when C# loads the C-Core, it throws an exception.

Long story short, the GRPC C-Core lib in the nuget package seems to be only available for amd64, which also readelf indicates. (Same for the protoc compiler)

### Describe the solution you'd like
ARM versions of C-Core libraries in nuget packages 
(BTW: I am about to investigate the same with other languages such as python and go, where GRPC packages are available via package managers - Can anybody tell me whether I will face the same problem?)

### Describe alternatives you've considered
Build C-Core for arm64 manually and do not use "easy installation" via packages.

### Additional context
I am sorry for the long text, hope the context helps though.

(I will attend the GRPC conf.)

**Comment by nicolasnoble** [comment]
We're most likely not to distribute arm binary packages, due to the lack of proper arm infrastructure to create and test these builds.

**Comment by jtattermusch** [comment]
Yes, we're not planning to ship prebuilt arm native libraries in our nuget because:
1. we don't have the proper infrastructure to build and test this
2. the Grpc.Core nuget is already pretty large in size (due to different builds of grpc_csharp_ext that need to be included) and we don't have a good way of splitting the package per-platform.

**Comment by vielmetti** [comment]
@nicolasnoble @jtattermusch 

What infrastructure would you need to build and test this?

I ask because @worksonarm has been providing infrastructure to Google-sponsored projects, including Go and Bazel, and would be happy to entertain any reasonable request for modern arm64 server hardware to support the build and test cycle.

**Comment by jsmouret** [comment]
If that helps, to get it working on a raspberry pi, here are the steps:

```shell
# add debian backports to get a fresh grpc plugin
echo "deb http://deb.debian.org/debian buster-backports main" | sudo tee /etc/apt/sources.list.d/backports.list

# add debian keys to use backports
sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 7638D0442B90D010
sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 04EE7237B7D453EC

# install protoc+plugin
sudo apt install -t buster-backports protobuf-compiler-grpc

# tell dotnet to use the local protoc
export PROTOBUF_PROTOC=/usr/bin/protoc

# trick dotnet to use the local grpc plugin as there are no environment variable to help
ln -s /usr/bin ~/.nuget/packages/grpc.tools/2.29.0/tools/linux_

# profit!
dotnet run
```

**Comment by km-hussain-in** [comment]
@jsmouret your solution does not work on 64-bit raspbian 

E: The value 'buster-backports' is invalid for APT::Default-Release as such a release is not available in the sources

**Comment by jsmouret** [comment]
Can you try an `apt-get update` ?

**Comment by jrdeboer** [comment]
@jsmouret I can't get your solution to work. I've done all the steps but the last one doesn't seem to work.
When I try to link the Nuget location I get this error: '/c/Users/MyUser/.nuget/packages/grpc.tools/2.29.0/tools/linux_': No such file or directory
Do you have any ideas?

**Comment by rubit0** [comment]
I'm currently also facing the same problem where I need an ARM64 version for HoloLens because Unity Reflect has dependency to GRPC.Core.
Is it possible to build my own ARM64 build of GRPC.Core?

**Comment by jsmouret** [comment]
@jrdeboer maybe the last step did not work because the Grpc.Tools version is different. you can try:
```
ln -s /usr/bin ~/.nuget/packages/grpc.tools/2.37.0/tools/linux_
```

**Comment by jtattermusch** [comment]
Just a note:
Thanks to https://github.com/grpc/grpc/pull/25717, Grpc.Core and Grpc.Tools now support linux arm64.

The support has already been released as part of the 2.37.0: https://www.nuget.org/packages/Grpc.Core/2.37.0
